mono/metadata/class.c

```diff
@@ -627,6 +627,289 @@ mono_type_get_name (MonoType *type)
+			}
+
+		}
+		else {
+			bufferIter = mono_identifier_escape_type_append(bufferIter, klass->name);
+		}
+
+		if (!is_recursed) {
+			if (bufferIter != bufferPtr)
+				bufferIter = flush_name_buffer(bufferPtr, nameChunkReport, userData);
+
+			if (mono_class_is_ginst(klass)) {
+				MonoGenericClass *gclass = mono_class_get_generic_class(klass);
+				MonoGenericInst *inst = gclass->context.class_inst;
+				MonoTypeNameFormat nested_format;
+				int i;
+
+				nested_format = format == MONO_TYPE_NAME_FORMAT_FULL_NAME ? MONO_TYPE_NAME_FORMAT_ASSEMBLY_QUALIFIED : format;
+
+				if (format == MONO_TYPE_NAME_FORMAT_IL)
+					*bufferIter++ = '<';
+				else
+					*bufferIter++ = '[';
+
+				for (i = 0; i < inst->type_argc; i++) {
+					MonoType *t = inst->type_argv[i];
+
+					if (i)
+						*bufferIter++ = ',';
+					if ((nested_format == MONO_TYPE_NAME_FORMAT_ASSEMBLY_QUALIFIED) &&
+						(t->type != MONO_TYPE_VAR) && (type->type != MONO_TYPE_MVAR))
+						*bufferIter++ = '[';
+
+					//flush the buffer before recursing
+					bufferIter = flush_name_buffer(bufferPtr, nameChunkReport, userData);
+					mono_type_get_name_foreach_name_chunk_recurse(inst->type_argv[i], FALSE, nested_format, nameChunkReport, userData);
+
+					if ((nested_format == MONO_TYPE_NAME_FORMAT_ASSEMBLY_QUALIFIED) &&
+						(t->type != MONO_TYPE_VAR) && (type->type != MONO_TYPE_MVAR))
+						*bufferIter++ = ']';
+				}
+				if (format == MONO_TYPE_NAME_FORMAT_IL)
+					*bufferIter++ = '>';
+				else
+					*bufferIter++ = ']';
+			}
+			else if (mono_class_is_gtd(klass) &&
+				(format != MONO_TYPE_NAME_FORMAT_FULL_NAME) &&
+				(format != MONO_TYPE_NAME_FORMAT_ASSEMBLY_QUALIFIED)) {
+				int i;
+
+				if (format == MONO_TYPE_NAME_FORMAT_IL)
+					*bufferIter++ = '<';
+				else
+					*bufferIter++ = '[';
+
+				bufferIter = flush_name_buffer(bufferPtr, nameChunkReport, userData);
+
+				for (i = 0; i < mono_class_get_generic_container(klass)->type_argc; i++) {
+					if (i)
+						nameChunkReport(",", userData);
+					nameChunkReport(mono_generic_container_get_param_info(mono_class_get_generic_container(klass), i)->name, userData);
+				}
+				if (format == MONO_TYPE_NAME_FORMAT_IL)
+					*bufferIter++ = '>';
+				else
+					*bufferIter++ = ']';
+			}
+
+			if (type->byref)
+				*bufferIter++ = '&';
+		}
+
+		if (bufferPtr != bufferIter)
+			nameChunkReport(bufferPtr, userData);
+
+		if ((format == MONO_TYPE_NAME_FORMAT_ASSEMBLY_QUALIFIED) &&
+			(type->type != MONO_TYPE_VAR) && (type->type != MONO_TYPE_MVAR)) {
+			MonoImage *klassImg = mono_class_get_image(klass);
+			char *imgName = mono_image_get_name(klassImg);
+			nameChunkReport(imgName, userData);
+		}
+		break;
+	}
+}
+
+/**
+ * mono_type_get_name_chunked:
+ * \param type a type
+ * \reports chunks null terminated of a type's name via a callback.
+ */
+void mono_type_get_name_chunked(MonoType * type, uint32_t format, void(*namePieceReportCallback)(void *data, void *userData), void * userData)
```

Comment:
Does all this need to be in here versus a Unity specific file? Thinking of pain when merging upstream.

Reply:
you are right, I was originally under the impression that this needed to be in here. fixing